app_flutter/lib/service/appengine_cocoon.dart

```diff
@@ -100,6 +101,50 @@ class AppEngineCocoonService implements CocoonService {
     return response.statusCode == HttpStatus.ok;
   }
 
+  /// Downloads the log for [task] and writes the bytes to the local storage of
+  /// the current device. Returns true if write was successful, and false if
+  /// there was a failure. Only works on the web platform.
+  ///
+  /// Steps to download the log:
+  /// 1. Make an authenticated request to the Cocoon backend for the task log.
+  /// 2. Encode the log to be uri safe.
+  /// 3. Create an anchor element with the encoded log data for downloading.
+  /// 4. Click the anchor element to trigger the browser to download the log.
+  @override
+  Future<bool> downloadLog(Task task, String idToken) async {
+    assert(task != null);
+    assert(idToken != null);
+
+    /// This is a web only solution for downloading logs.
+    if (!kIsWeb) {
+      return false;
+    }
+
+    final String getTaskLogUrl =
+        _apiEndpoint('/api/get-log?ownerKey=${task.key.child.name}');
+
+    /// This API endpoint returns the log in the body of the response.
+    final http.Response response = await _client.get(
+      getTaskLogUrl,
+      headers: <String, String>{
+        'X-Flutter-IdToken': idToken,
+      },
+    );
+
+    if (response.statusCode != HttpStatus.ok) {
+      print(response.body);
+      return false;
+    }
+
+    final String logEncodedContent = Uri.encodeComponent(response.body);
```

Comment:
Is this slow on larger log messages? Is there a way we could use built-in browser functionality instead?

Reply:
Good catch! This actually crashes on larger log messages. See [microbenchmark_ios](https://flutter-dashboard.appspot.com/api/get-log?ownerKey=ahFmbHV0dGVyLWRhc2hib2FyZHJYCxIJQ2hlY2tsaXN0IjhmbHV0dGVyL2ZsdXR0ZXIvNDM5ZDc5ODcxMjU1MDc2NTkwYjQ2YWUyMjA5OWEwZTIwZjA2YzA5YwwLEgRUYXNrGICA0PDOkekJDA). The browser returns `Failed - Network error` when trying to store that much data in `AnchorElement`. Granted, that link still crashes my browser when I open it. I haven't been able to locate a size limit for this current approach.

An alternative is to just write the idToken to a cookie for now, and open the log url. This would be the same as v1. The original reason for not doing this was because we didn't have storing the devicelab logs in GCS prioritized until now. With https://github.com/flutter/flutter/issues/43124 being worked on, we can just redirect users to the Google Cloud Log Viewer link with the GCS url passed in which is an ideal solution when that work is done.

Do you know of any other browser built in way for triggering downloads? There isn't much documentation on it, so I was only able to find the trick with using an anchor element to download the downloaded data.

Reply:
For reference: The largest log I was able to download (just randomly trying) was 300kB.

Reply:
I think you can put a `download` attribute on an anchor element and that will trigger a download when you click on it. i don't know if that will fix the crashing.

Reply:
That's what the current approach did :)

I have rewritten it to write a cookie first, then just download via the log url. Before it was downloading the log data, then encoding it for the anchor element.

This approach works with the logs that crash chrome in v1 (1.9 mb log referenced above works like a charm).

Just need to fix the imports so it can run the tests :)

Reply:
Sounds good!